### DISCUSSION TITLE
Multiple Gateways?

### DISCUSSION THREAD

**Discussion by akropp** [jasonacox/Powerwall-Dashboard#62] [Q&A]
My system is set up with two gateways, each being fed by a powerwall+ inverter.  It seems that you can handle multiple PW+ on one system, but is there anything already in place to handle multiple gateways?  I was thinking I could just run two proxies and then have the data merged into influxdb, but don't know if I can do that in a way that will flow through everything else without breaking it.  Any thoughts?  I'm happy to be a guinea pig/dig into the code if necessary.

**Comment by jasonacox** [comment]
Hi @akropp !  Thanks for opening this.  Does each gateway have a separate IP address?  If so, my recommendation would be to spin up two pypowerwall instances (containers) on separate ports (say 8676 for pypowerwall2), one for each gateway, then edit the telegraf config to have a separate section (inputs.http) for the 2nd gateway and add tag.  It would still require updating the Grafana dashboard.json to accommodate, but at least all the data would be in the same spot for both gateways.  If you are able to get this to work, we should add this as an optional configuration to the project!

I haven't tried it but I think something like this for `telegraf.conf`:

```
[[inputs.http]]
	urls = [
		"http://pypowerwall:8675/aggregates",
		"http://pypowerwall:8675/soe",
		"http://pypowerwall:8675/strings",
		"http://pypowerwall:8675/temps/pw",
		"http://pypowerwall:8675/freq",
   		"http://pypowerwall:8675/pod"
	]
	method = "GET"
	insecure_skip_verify = true
	timeout = "4s"
	data_format = "json"
	[inputs.http.tags]
	host="pypowerwall"

	
[[inputs.http]]
	urls = [
		"http://pypowerwall2:8676/aggregates",
		"http://pypowerwall2:8676/soe",
		"http://pypowerwall2:8676/strings",
		"http://pypowerwall2:8676/temps/pw",
		"http://pypowerwall2:8676/freq",
    		"http://pypowerwall2:8676/pod"
	]
	method = "GET"
	insecure_skip_verify = true
	timeout = "4s"
	data_format = "json"
	[inputs.http.tags]
	host="pypowerwall2"
```

**Reply by nberardi** [reply]
Has anyone tried this? I have a 400 A service and it requires two powerwalls and two gateways.

**Reply by jasonacox** [reply]
Are the gateways registering as two different Tesla installations or can you see both gateways (and combined power metrics) in the Tesla app?  I'm curious as to how Tesla manages and represents it and seeing if we can mimic that.

**Reply by nberardi** [reply]
As far as I understand the setup, which I can confirm tomorrow after installation. Each gateway is registered as a different installation. I am not sure how the app combines the two yet.

**Reply by fasternfaster** [reply]
I currently only see one gateway in my tesla app but it is reporting combined kW numbers and what seems like the average battery charge number.  I have attempted to update the telegraf.conf file as you stated. I also added the following to powerwall.yml
```yml
 pypowerwall2:
        image: jasonacox/pypowerwall:latest
        container_name: pypowerwall2
        hostname: pypowerwall2
        restart: always
        user: "1000:1000"
        ports:
            - target: 8675
              published: 9675
              mode: host
        env_file:
            - pypowerwall2.env
 ```

The tags dont appear to work or i just dont understand grafana  at all. When i try to put the URL or the HOST in the WHERE clause, i get no data. And if i dont add the WHERE clause, the data for both systems/gateways/powerwalls just gets combined into one graph causing the line to jitter up down between the values.

**Reply by jasonacox** [reply]
Are you able to get to the pypowerwall2 container? (replace localhost with the IP of the host running the containers):
* http://localhost:9675/stats
* http://localhost:9675/example.html

If so, the next thing you need to do is to create the section in `telegraf.conf` to pull data from that pypowerwall2 instance:

```
[[inputs.http]]
	urls = [
		"http://pypowerwall2:9675/aggregates",
		"http://pypowerwall2:9675/soe",
		"http://pypowerwall2:9675/strings",
		"http://pypowerwall2:9675/temps/pw",
		"http://pypowerwall2:9675/freq",
    		"http://pypowerwall2:9675/pod"
	]
	method = "GET"
	insecure_skip_verify = true
	timeout = "4s"
	data_format = "json"
	[inputs.http.tags]
	host="pypowerwall2"
```

**Reply by fasternfaster** [reply]
<img width="1844" alt="Screenshot 2023-02-27 at 8 26 43 AM" src="https://user-images.githubusercontent.com/2657475/221575820-b9a1bf96-c76d-4d63-8054-12dd3cae352c.png">
you can see that the battery charge state in particular starts to jitter between two values after i made a change in the middle of this graph

**Reply by jasonacox** [reply]
Ah!  I see!  It is combining both and as @nberardi says, creating a mean.  You can see the charge level alternating between the two.

The dashboard queries would need to be rewritten to look at the "tags".   The data from one is "powerwall" and the other is "powerwall2" set by telegraf. I suspect there is a better way to sort he data.  I think I can simulate it by adding one to my test rig.  I'll see what I can figure out for you.

**Reply by fasternfaster** [reply]
thanks, i keep trying to add the tag, but when i select for it, i get no data, this makes me think that the tag is not being added. I also tried adding `key_tags=["PW1_temp"]` thinking i would need to list all the keys that need tags, but this just broke everything and nothing was working, so i reverted.. 

My eventual hope is to have the main graph SUM together the power from both systems and then i could add separate graphs below for each system/powerwall

**Reply by fasternfaster** [reply]
<img width="1846" alt="Screenshot 2023-03-08 at 8 22 55 AM" src="https://user-images.githubusercontent.com/2657475/223724460-feb7879b-0d18-49e8-b114-bc1c319466de.png">
I think i got it to a good working state finally!!  This is a 2 day view, you can see day 1 everything is mixed together and day 2 i have 2 separate lines for the two systems.

I ended up adding a second telegraf instance and pointing it to a "powerwall2" database name. Still just one influxdb. I had to update the sql scripts and all references of "powerwall" required duplicating some code. After this it was a matter of duplicating the grafana panels and simply changing the datasource and renaming them.

telegraf2.conf :
```
`[[outputs.influxdb]]
	urls = ["http://influxdb:8086"]
	database = "powerwall2"
	skip_database_creation = false
	retention_policy = "raw"

	
[[inputs.http]]
	urls = [
		"http://pypowerwall2:8675/aggregates",
		"http://pypowerwall2:8675/soe",
		"http://pypowerwall2:8675/strings",
		"http://pypowerwall2:8675/temps/pw",
		"http://pypowerwall2:8675/freq",
    "http://pypowerwall2:8675/pod"
	]
	method = "GET"
	insecure_skip_verify = true
	timeout = "4s"
	data_format = "json" `
```

powerwall.yml:
```
`    telegraf2:
        image: telegraf:latest
        container_name: telegraf2
        hostname: telegraf2
        restart: always
        user: "1000:1000"
        command: [
            "telegraf", 
            "--config",  
            "/etc/telegraf/telegraf.conf", 
            "--config-directory",
            "/etc/telegraf/telegraf.d"
        ]
        volumes:
            - type: bind
              source: ./telegraf2.conf
              target: /etc/telegraf/telegraf.conf
              read_only: true
            - type: bind
              source: ./telegraf.local
              target: /etc/telegraf/telegraf.d/local.conf
              read_only: true            
        depends_on:
            - influxdb
            - pypowerwall2 `
```

again i had to duplicate the code in all the .sql files also. 

Only thing i cannot do yet is compute the combined solar/house use/grid use of the 2 systems. Thats likely my lack of grafana knowledge.

**Reply by jasonacox** [reply]
Nice!  However, I think that separating the data into two databases is going to make it more difficult as you mention, to combine.  I couldn't find a quick fix for that either.  However, in the first version, the data should have separate host tags.  I'm not at a place where I can test much now but check out how to filter based on tags and "host".  Telegraf should be setting host tag based on "pypowerwall" or "pypowerwall2".  Or perhaps we have telegraf inject a tag?  I would have to test what we would add to WHERE clause.

**Reply by fasternfaster** [reply]
i tried with tags for days and never got anything to inject. Running a query with SELECT * from ... i got columns for the tags i tried to create, but the values were always 'null'.

**Reply by nberardi** [reply]
@fasternfaster I was in the same boat, however after moving to Flux on InfluxDB2, I pretty much started from scratch with built in support for the host tag. I am not that familiar with how Influx does CQ's, so I blame my ignorance.  

I assumed that the indexes that were getting created with the CQ's weren't aware of the host tag, however I also couldn't find a way to integrate it in either. Which is why I decided to start from scratch.

**Reply by jasonacox** [reply]
You're right!  I believe we would need to adjust the CQs to GROUP BY the host tag.

```sql
GROUP BY time(1h), month, year, host
```

I'll try to test out a few scenarios this weekend to see if I can solve this.

**Comment by akropp** [comment]
Yeh, each gateway shows up separately in the app (you swipe back and forth
between them).  There is no combined info.  I never ended up setting up the
two sessions, because I am mostly pulling the data in to Home Assistant and
keeping track of it there (although it doesn't track as many of the
stats).  I have pypowerwall set up pulling the stats from one still.
Probably would have been nice to have the second set up for historical
purposes.

--Adam

**Reply by jasonacox** [reply]
It's a bit disappointing that they don't have a combined view. 

I'm going to open an issue to help track this.  I suspect you and @nberardi are not alone in having this setup.  If you have time, would love to hear what you would like to see on the Dashboard.  I assume:
* For the primary "Live Monitoring" energy usage graph and the Power Meters, you would want combined data for solar, grid, home and powerwalls.
* For the power flow animation we would need two instances (side-by-side?)
* The detail graphs below would likely work best as separate breakouts for each gateway, but let me know what you think.

**Reply by jasonacox** [reply]
Issue: https://github.com/jasonacox/Powerwall-Dashboard/issues/148

**Reply by nberardi** [reply]
Anybody with 400 A service to their home and wanted a whole home coverage would be in the same boat as us.

**Comment by fasternfaster** [comment]
I have the same issue, also noticing a difference in reported powerwall charge level between the gateway page and the dashboard. could it be that with multiple powerwalls per gateway the wrong charge is being reported?

**Reply by jasonacox** [reply]
Hi @fasternfaster !  Can you expand a bit on that discrepancy (or perhaps paste a screenshot)?

**Reply by nberardi** [reply]
It takes the mean of the values entered into the database. Which usually means it is wrong when you have two systems. I spent a ton of time revamping the dashboard for two systems. Let me see if I can find the time to pull together a PR.

**Reply by fasternfaster** [reply]
i pasted one above, i made changes to both files (powerwall.yml and telegraf.conf) then i tried to update the simplest graphs (Inverter power and Powerwall temps)  They both currently show 2 lines since i have 4 total inverters and 4 powerwalls but each gateway only sees 2.  So each one is reporting data on #1 and #2 i guess? When i tried to edit the graph query with a WHERE clause the line disappeared.  I thought i might need an update so i deleted and recreated the telegraph container. When i put the WHERE clause in, grafana DOES suggest correct values for HOST (pypowerwall and pypowerwall2) and it also suggests correct values for URL. so that made me think i was on the right path, but selecting them returns nothing.
<img width="659" alt="Screenshot 2023-02-27 at 8 50 44 AM" src="https://user-images.githubusercontent.com/2657475/221581229-bd49e120-ab50-4fa5-b693-6e77cde3bf20.png">

**Reply by nberardi** [reply]
From what's its worth, I wasn't able to get the dashboard working with InfluxQL and InfluxDB1.x.  This was probably due to a knowledge gap on my part with InfluxQL.  However I was able to get a multi-gateway setup working in InfluxDB2.x.  https://github.com/nberardi/Powerwall-Dashboard/commit/d0da1cc61039b007e39ba5e14783dd97117f91b5

**Reply by fasternfaster** [reply]
this didn't work for me unfortunately. i think i got everything up and running with no errors from telegraf but my dashboard stopped updating completely.  The dashboard you had added was no loading for me due to an error 'Datasource cNwAjrt4k was not found'

**Reply by swgamerx** [reply]
Any updates on this? I attempted to create two completely separate running instances of all the different dockers but It's having fits over the port numbers even though I changed them in the powerwall.yml file. Is the best approach to have two dockers running pypowerwall and two databases in a single influxdb instance?
